dbms/src/Functions/FunctionsArithmetic.h

```diff
@@ -873,8 +883,65 @@ class FunctionBinaryArithmetic : public IFunction
         return type_res;
     }
 
+    bool isAggregateMultiply(const DataTypePtr & type0, const DataTypePtr & type1) const
+    {
+        if constexpr (!std::is_same_v<Op<UInt8, UInt8>, MultiplyImpl<UInt8, UInt8>>)
+            return false;
+        auto is_uint_type = [](const DataTypePtr & type) 
+        {
+            return checkDataType<DataTypeUInt8>(type.get()) || checkDataType<DataTypeUInt16>(type.get())
+                || checkDataType<DataTypeUInt32>(type.get()) || checkDataType<DataTypeUInt64>(type.get());
+        };
+        return ((checkDataType<DataTypeAggregateFunction>(type0.get()) && is_uint_type(type1))
+            || (is_uint_type(type0) && checkDataType<DataTypeAggregateFunction>(type1.get())));
+    }
+
     void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) override
     {
+        /// Special case when multiply aggregate function state
+        if (isAggregateMultiply(block.getByPosition(arguments[0]).type, block.getByPosition(arguments[1]).type))
+        {
+            ColumnNumbers new_arguments = arguments;
+            if (checkDataType<DataTypeAggregateFunction>(block.getByPosition(new_arguments[1]).type.get()))
+                std::swap(new_arguments[0], new_arguments[1]);
+                
+            const ColumnAggregateFunction * column = typeid_cast<const ColumnAggregateFunction *>(block.getByPosition(new_arguments[0]).column.get());
+            IAggregateFunction * function = column->getAggregateFunction().get();
+
+            MutableColumnPtr current = column->cloneEmpty();
+            auto arena = std::make_shared<Arena>();
+            auto & res = typeid_cast<ColumnAggregateFunction &>(*current);
+            res.reserve(input_rows_count);
+            auto & vec_to = res.getData();
+            const auto & vec_from = column->getData();
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+                res.insertDefault();
+
+            UInt64 m = block.getByPosition(new_arguments[1]).column->getUInt(0);
+
+            /// We use exponentiation by squaring algorithm to perform multiplying aggregate states by N in O(log(N)) operations
+            /// https://en.wikipedia.org/wiki/Exponentiation_by_squaring
+            while (m)
+            {
+                if (m % 2)
+                {
+                    for (size_t i = 0; i < input_rows_count; ++i)
+                        function->merge(vec_to[i], vec_from[i], arena.get());
+                    --m;
+                }
+                else
+                {
+                    for (size_t i = 0; i < input_rows_count; ++i)
+                        function->merge(vec_from[i], vec_from[i], arena.get());
```

Comment:
Are we modifying source column? It's not permitted.

(It should not even be able to compile, but actually you get AggregateDataPtr instead of ConstAggregateDataPtr from const ColumnAggregateFunction.)